R/guide-.R

```diff
@@ -0,0 +1,344 @@
+#' Guide constructor
+#'
+#' A constructor function for guides, which performs some standard compatability
+#' checks between the guide and provided arguments.
+#'
+#' @param ... Named arguments that match the parameters of `super$params` or
+#'   the theme elements in `super$elements`.
+#' @param available_aes A vector of character strings listing the aesthetics
+#'   for which the guide can be drawn.
+#' @param super The super class to use for the constructed guide. Should be a
+#'   Guide class object.
+#'
+#' @return A `Guide` ggproto object.
+#' @keywords internal
+#' @export
+new_guide <- function(..., available_aes = "any", super) {
+
+  pf <- parent.frame()
+  super <- check_subclass(super, "Guide", env = pf)
+
+  args <- list2(...)
+
+  # Set parameters
+  param_names <- names(super$params)
+  params <- intersect(names(args), param_names)
+  params <- defaults(args[params], super$params)
+
+  # Set elements
+  elems_names <- names(super$elements)
+  elems  <- intersect(names(args), elems_names)
+  elems  <- defaults(args[elems], super$elements)
+
+  # Warn about extra arguments
+  extra_args <- setdiff(names(args), union(param_names, elems_names))
+  if (length(extra_args) > 0) {
+    cli::cli_warn(paste0(
+      "Ignoring unknown {cli::qty(extra_args)} argument{?s} to ",
+      "{.fn {snake_class(super)}}: {.arg {extra_args}}."
+    ))
+  }
+
+  # Stop when some required parameters are missing.
+  # This should only happen with mis-constructed guides
+  required_params <- names(Guide$params)
+  missing_params  <- setdiff(required_params, names(params))
+  if (length(missing_params) > 0) {
+    cli::cli_abort(paste0(
+      "The following parameter{?s} {?is/are} required for setting up a guide, ",
+      "but {?is/are} missing: {.field {missing_params}}"
+    ))
+  }
+
+  # Ensure 'order' is length 1 integer
+  params$order <- vec_cast(params$order, 0L, x_arg = "order", call = pf)
+  vec_assert(params$order, 0L, size = 1L, arg = "order", call = pf)
+
+  ggproto(
+    NULL, super,
+    params   = params,
+    elements = elems,
+    available_aes = available_aes
+  )
```

Comment:
The guide constructor never updates the class with a user's input. It takes the user's input and fuses it to the guide's defaults, which lives on in a child instead of modifying the parent.